src/test_report.py

```diff
@@ -108,39 +92,18 @@ def _get_results_data(self, root_node):
             'groups': groups,
         }
 
-    def _create_test_report(self, root_node):
-        results = self._get_results_data(root_node)
-        template_env = jinja2.Environment(
-                            loader=jinja2.FileSystemLoader("./config/reports/")
-                        )
-        template = template_env.get_template("test-report.jinja2")
-        revision = root_node['revision']
-        stats = results['stats']
-        groups = results['groups']
-        subject = (f"\
-{revision['tree']}/{revision['branch']} {revision['describe']}: \
-{stats['total']} runs {stats['failures']} failures")
-        content = template.render(
-            subject=subject, root=root_node, groups=groups
-        )
-        return content, subject
-
-    def _send_email(self, content, subject):
-        smtp = self._smtp_connect()
-        if smtp:
-            email_msg = self._create_email(
-                self._email_send_to, self._email_send_from,
-                subject, content
-            )
-            smtp.send_message(email_msg)
-            smtp.quit()
-
-    def _make_report(self, root_node, dump=True, send=False):
+    def _get_report(self, root_node, dump=True, send=False):
         content, subject = self._create_test_report(root_node)
         if dump:
             print(content, flush=True)
         if send:
-            self._send_email(content, subject)
+            try:
+                email_msg = self._email_sender.create_email(
+                    subject, content
+                )
+                self._email_sender.send_email(email_msg)
+            except Exception as err:
+                self._logger.log_message(logging.ERROR, err)
 
     def run_loop(self):
```

Comment:
I'm not entirely convinced that `run_loop` logic belongs to the TestReport. For me it belongs rather to the `class cmd_run`. Is there a reason why it's here?

Reply:
The reason being is to align the class with the pattern of other client scripts.
Also, I created two different classes to run the service in a loop and just for a single node by the `loop` and `run` commands respectively.

Reply:
@JenySadadia 
I get your point, so if it is a pattern in other scripts I guess we should follow it. Creating 2 classes just to run single and loop versions seems a bit artificial.
I guess we should get back to the previous implementation with `run_loop()` and `run()`, though the design doesn't seem optimal. Can we do it?

Reply:
Can you please elaborate on the comment?
In my opinion, It is easier to maintain two different classes (having two different commands to serve) when the code grows gradually. I thought we did the same thing for `runner` as well.

Reply:
Ok, I get your point, however I think that implementation is not quite right. `TestReportLoop` and `TestReportSingle` classes have different purpose than `TestReport`. They are a kind of "runners" of the logic provided by `TestReport`, hence they shouldn't inherit from `TestReport`, they should rather take an instance of `TestReport` while they're constructed, store it as an attribute and use it in the `.run()` method.

Maybe later we can also create a separate `Runner` class that could be use to run code in the other pipeline components, but that's a follow up.

Reply:
If I don't inherit the `TestReport` class and use its instance instead, I need to use protected members like `_db` and `_logger` from `TestReportLoop` and `TestReportSingle` classes. That will not be a good coding practice.
Rather, I chose to go with the old implementation of two separate methods but renamed them to `loop` and `run`.